framework/src/org/checkerframework/common/value/RangeOrListOfValues.java

```diff
@@ -0,0 +1,95 @@
+import java.util.Collections;
+import java.util.List;
+import javax.lang.model.element.AnnotationMirror;
+import org.checkerframework.common.value.qual.ArrayLen;
+import org.checkerframework.common.value.qual.ArrayLenRange;
+import org.checkerframework.common.value.qual.IntVal;
+import org.checkerframework.common.value.util.Range;
+import org.checkerframework.framework.type.AnnotatedTypeMirror;
+
+/**
+ * An abstraction that can be either a range or a list of values that could come from an {@link
+ * ArrayLen} or {@link IntVal}. This abstraction reduces the number of cases that {@link
+ * org.checkerframework.common.value.ValueAnnotatedTypeFactory.ValueTreeAnnotator#handleInitializers(List,
+ * AnnotatedTypeMirror.AnnotatedArrayType)} and {@link
+ * org.checkerframework.common.value.ValueAnnotatedTypeFactory.ValueTreeAnnotator#handleDimensions(List,
+ * AnnotatedTypeMirror.AnnotatedArrayType)} must handle.
+ */
+class RangeOrListOfValues {
+    Range range;
+    List<Integer> values;
+    boolean isRange;
+
+    public RangeOrListOfValues() {
+        this.values = new ArrayList<>();
+        isRange = false;
+    }
+
+    public RangeOrListOfValues(List<Integer> values) {
+        this.values = new ArrayList<>();
+        isRange = false;
+        addAll(values);
+    }
+
+    public RangeOrListOfValues(Range range) {
+        this.range = range;
+        isRange = true;
+    }
+
+    /**
+     * If this is not a range, adds all members of newValues to the list. Otherwise, extends the
+     * range as appropriate based on the max and min of newValues. If adding newValues to a
+     * non-range would cause the list to become too large, converts this into a range.
+     */
+    public void addAll(List<Integer> newValues) {
+        if (isRange) {
+            Range newValueRange = new Range(Collections.min(newValues), Collections.max(newValues));
+            range = range.union(newValueRange);
+        } else {
+            for (Integer i : newValues) {
+                if (!values.contains(i)) {
+                    values.add(i);
+                }
+            }
+            if (values.size() > ValueAnnotatedTypeFactory.MAX_VALUES) {
+                convertToRange();
+            }
+        }
+    }
+
+    /**
+     * Produces the most precise annotation that captures the information stored in this
+     * RangeOrListofValues. The result is either a {@link ArrayLen} or a {@link ArrayLenRange}.
+     */
+    public AnnotationMirror createAnnotation(ValueAnnotatedTypeFactory atypefactory) {
+        if (isRange) {
+            return atypefactory.createArrayLenRangeAnnotation(range);
+        } else {
+            return atypefactory.createArrayLenAnnotation(values);
+        }
+    }
+
+    // To be called before addAll
+    public static List<Integer> convertLongsToInts(List<Long> newValues) {
+        List<Integer> result = new ArrayList<>(newValues.size());
+        for (Long l : newValues) {
+            result.add(l.intValue());
```

Comment:
Why is it safe to ignore overflow here?  If l is too big, then the ArrayLen* annotation will end up with negative array lengths.